### DISCUSSION TITLE
How do I change css properties of a class

### DISCUSSION THREAD

**Discussion by conormdowney** [adobe/react-spectrum#1933] [General]
Im wondering if someone can help me with how to change css properties of spectrum classes. 

The example ive been using is, for the class _spectrum-Textfield-input_61339 Im trying to update the font-family to Times New Roman (not an actual use case, just for testing). 

Im not very well versed in css, so please bear with me.

i made a css class and i import it in my react app. That doesnt do anything. If I make the property !important it works. 

It seems like the spectrum css is loaded in after my one. Is my issue then just a build problem, in that i should be loading my css after spectrums? 
Or is there a method for doing this?

**Comment by snowystinger** [comment]
So there are a few things to consider here.
- Please don't override things if they are hard to override, they typically are that way because we're trying to enforce a Design System.
  - To this note, we use CSS Modules to make it harder to override, so if you're using a classname such as _spectrum-Textfield-input_61339, that hash on the end will change anytime we update that file, meaning that you could be using a broken selector
- For most things you need to style within the context of the Design System, check https://react-spectrum.adobe.com/react-spectrum/styling.html
- You're running into a specificity issue, I recommend reading up on this topic, it's not always obvious https://css-tricks.com/specifics-on-css-specificity/ with the caveat that maybe you shouldn't be overriding that property if it's hard to do so

If you are a non-adobe employee, we also have stately and aria hooks that power all of our components behaviors, you could provide your own rendering on top of that if you need to make changes that our Design System doesn't allow.

**Reply by conormdowney** [reply]
Thank you for the quick reply. 

The react aria hooks, if im understanding this correctly, I can use useSearchField (its a SearchField i am testing on) and give it some value for the font in the props and then use the label and input props fields to build my own component that will still have all the other default styling values with just the one i entered in props different?

**Reply by snowystinger** [reply]
hooks contain no knowledge of styles or dom structure or anything, you'd need to recreate this file essentially https://github.com/adobe/react-spectrum/blob/main/packages/@react-spectrum/searchfield/src/SearchField.tsx
but change all the classes to your classes with your styles, specifically the inputClassName that TextFieldBase accepts
That might be enough, but there may be more work required.
You may also be able to copy our CSS for the SearchField and use that as the starting point for your customized SearchField.

**Reply by conormdowney** [reply]
Okay. thanks for all the info. It was very helpful

**Reply by snowystinger** [reply]
I think I might be of more help if you actually explain what you're wanting to do. You said changing the font wasn't actually it, and font is a really difficult one to deal with because it's so core to the Design System. So what is it you're actually trying to accomplish?

**Reply by conormdowney** [reply]
Basically our design team were just wondering if we could change styles from spectrums default settings. Font was mentioned, we want to use Roboto everywhere (I think) but it was more of a general question of can we change default styles to be what our design team wants. 

EDIT: sorry. should be clearer. when i say change I mean override with new values

**Reply by snowystinger** [reply]
Ah, you really want to be looking more at theme overriding i think, there are css variables for most everything, you may be able to change font that way instead.
You can see how that works here https://github.com/adobe/react-spectrum/blob/main/packages/%40react-spectrum/provider/stories/Provider.stories.tsx#L88
To find the variables though, you'll likely have to inspect around in the dom, we don't have a good place to list all of them yet

**Reply by conormdowney** [reply]
Cool. Just had a quick look. Forgive my massive ignorance, im not sure what im looking at through that link but i do see a bunch of variables defined in css classes here:

https://github.com/adobe/react-spectrum/tree/main/packages/%40adobe/spectrum-css-temp/vars

so if i want to make my own default theme i would use a subset of them? Would i need to make it an exhaustive list of all the variables i want to use? In which case, ones not defined will not have a value or will they default to the ones defined though that link?

**Reply by snowystinger** [reply]
well so if you're looking at changing the font, you'd really be interested in this file https://github.com/adobe/react-spectrum/blob/main/packages/@adobe/spectrum-css-temp/components/commons/fonts.css
As for the link I provided, it shows how to use a css file to create a custom theme that overrides a couple specific css variables (such as the ones in this fonts.css file) and it shows how to pass that information to the React Spectrum Provider

You can see that story in action if you build React Spectrum locally, you can follow the guide here: https://react-spectrum.adobe.com/contribute.html#developing

**Reply by conormdowney** [reply]
Okay. cool. Im gonna make a run at this later today. ill prob be back with more silly questions then. Thanks

**Reply by conormdowney** [reply]
I might be asking a real dumb question here but i still need to ask it.

in the link you gave me path is: @react-spectrum/provider/stories/Provider.stories.tsx 

in my node mdules I dont see that path available. Under @adobe/react-spectrum there is no provider folder. Theres a module.js and there is a reference to Provider in there:

export {Provider} from '@react-spectrum/provider'

When I open that I see the following file:

```
import { ProviderContext, ProviderProps } from "@react-types/provider";
import React from "react";
/**
 * Provider is the container for all React Spectrum applications.
 * It defines the theme, locale, and other application level settings,
 * and can also be used to provide common properties to a group of components.
 */
export let Provider: React.ForwardRefExoticComponent<ProviderProps & React.RefAttributes<import("@react-types/shared").DOMRefValue<HTMLDivElement>>>;
export function useProvider(): ProviderContext;
export function useProviderProps<T>(props: T): T;

//# sourceMappingURL=types.d.ts.map

```

I went looking because in the file though the link you gave me there is an import:

import scaleLarge from '@adobe/spectrum-css-temp/vars/spectrum-large.css';

I cant do that in my project. I went looking through the node modules but i dont seem to have spectrum-css-temp. Also the temp in that folder name throws me. Is that something I should actually be using?

**Reply by snowystinger** [reply]
@adobe/react-spectrum is just a package that re-exports a lot of other packages to make it easy to only have one version of everything and know that the versions you have will all work together.

You don't need to import our css files because you can just use the default theme we also include in that package.
```
<Provider with defaultTheme>
  <Provider with customTheme changing just pieces of the overall>
  </Provider>
</Provider>
```

The storybook somewhat hides this, but every story in Provider.stories.tsx has a `Provider` rendered around it, thats why you don't see the defaultTheme imported or used anywhere in that file. As for medium and large, you could probably pass an empty object.

**Reply by conormdowney** [reply]
Okay. didnt realise you meant have a second Provider under the first that could override the values. 
As soon as I figure out how to turn on css modules in my project I will give this a shot.

**Reply by snowystinger** [reply]
Css modules is just a mapping of class names to hashed class names. You can just pass an object with keys that match their values. You aren't required to use css modules

**Reply by conormdowney** [reply]
Okay, making progress (i think). not using css modules, instead I just create a map as below.

```
import {Theme} from '@react-types/provider';

const overrideGlobal = {"--spectrum-alias-font-size-default": "70px"}

console.log(global)
export let theme: Theme = {
  global: overrideGlobal
};
```

But I get to a line of code where this causes a problem.
```

let themeKey = Object.keys(theme[colorScheme])[0];
  let scaleKey = Object.keys(theme[scale])[0];
```
Its in module.js (im not sure how this is built so that might now mean anything to you) but its the Provider Wrapper code. It seems I need an entry in my theme for the selected colorScheme and scale. but without css modules to do as its done in the link you provided, can I do this?

**Reply by snowystinger** [reply]
Try console logging out what the theme object is in the storybook, that should give you the clue you need. Sorry, on my phone

**Reply by snowystinger** [reply]
were you able to figure it out?
you should be able to have a css file like https://github.com/adobe/react-spectrum/blob/main/packages/%40react-spectrum/provider/stories/custom-theme.css
where you can define
```
.myOverrideVars {
--spectrum-alias-font-size-default: 70px;
}
```

then in the js file, import that css file, but since you aren't using CSS modules, create an object like so
```
const THEME = {  light: {myOverrideVars: 'myOverrideVars'},  medium: scaleMedium,  large: scaleLarge};
```

The Provider will take that first (and only) entry in the object and use it as the className, as shown here https://github.com/adobe/react-spectrum/blob/main/packages/%40react-spectrum/provider/src/Provider.tsx#L135

**Reply by conormdowney** [reply]
Sorry. it was a long weekend here and i was away from my laptop for the weekend.

Isnt CSS modules used though to create the scaleMedium and scaleLarge objects?
cos this is pretty much what ive done except i dont have those two objects

**Reply by snowystinger** [reply]
oh, yeah, so since those are already defined in the Provider wrapping yours, you could just do
`const THEME = {  light: {myOverrideVars: 'myOverrideVars'},  medium: {},  large: {}};`
or if you want the font size to change with scale, you can give an empty object to light instead and instead put that kind of an object into medium and large, the key shouldn't matter, but the value in the object should match your classname where you have defined the variable

one thing i'm not sure on is if you can give an empty object, you might need to send in `{foo: 'foo'}` or something in whichever one(s) you aren't using

**Reply by conormdowney** [reply]
So Im not getting an error anymore but it doesnt seem to use my custom defined var but im configuring it wrong. the value in the const that im adding to the theme was 

`const overrideGlobal = {"--spectrum-alias-font-size-default": "70px"}`

that wont work cos its defined in this css file (https://github.com/adobe/react-spectrum/blob/main/packages/%40adobe/spectrum-css-temp/vars/spectrum-global.css) under a class. 

so I chanced my arm and tried 

```const overrideGlobal = {".spectrum--medium,.spectrum--large" : `{--spectrum-alias-font-size-default": "70px"}`}```

but that doesnt work. so how do I define a variable value inline?

**Reply by snowystinger** [reply]
you don't define it inline, or if you do, you do it in a style tag or in a style object
```
// this is what importing css files into js will do, but it'll add it to the head of the document, whereas you could technically render this anywhere
<style>
.myOverrideVars {
  --spectrum-alias-font-size-default: 70px;
}
</style>
<div style={{'--spectrum-alias-font-size-default': '70px'}} >...</div>
```
I don't recommend the two options above, the div one won't propagate through a portal and the style one would be odd because it's not in the head with the rest of the css.

the object you pass to a Provider should have a classname only in it, no styles, and it should match the classname in the css file is where you do define the variable

**Reply by conormdowney** [reply]
Thats what I tried when i chanced my arm:

```const overrideGlobal = {".spectrum--medium,.spectrum--large" : `{--spectrum-alias-font-size-default": "70px"}`}```

I got that class name(s) from the spectrum_global.css file. so am i on the right track there? am I setting the value of the object incorrect?

**Reply by snowystinger** [reply]
I think this will clear things up https://codesandbox.io/s/mutable-moon-u8d8j?file=/src/App.js

**Reply by conormdowney** [reply]
That is awesome. Thank you so much. I was totally misreading. I thought the myOverrideVars was just you using an example class like TestClass or something. Thanks so much for your time and patience. You're a champ!!
